File: FairphonePeaceOfMind/src/main/java/ca/mudar/fairphone/peaceofmind/utils/TimeHelper.java
/*
 * Copyright (C) 2013 Mudar Noufal, PeaceOfMind+
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package ca.mudar.fairphone.peaceofmind.utils;

import android.content.res.Resources;
import android.text.format.DateUtils;

import java.util.Calendar;
import java.util.Locale;

import ca.mudar.fairphone.peaceofmind.Const;
import ca.mudar.fairphone.peaceofmind.R;

public class TimeHelper {
    private static final String TAG = "TimeHelper";

    public static String[] generateStringTimeFromMillis(long timePast, boolean reset, Resources res) {
        int hours = 0;
        int minutes = 0;
        if (!reset) {
            hours = (int) (timePast / Const.HOUR);
            int timeInMinutes = (int) (timePast - hours * Const.HOUR);

            if (hours == 0) {
                minutes = timeInMinutes - Const.MINUTE > 0 ? timeInMinutes / Const.MINUTE : 1;
            } else {
                minutes = timeInMinutes / Const.MINUTE;
            }
        }

        final String timeStr = String.format(Locale.US, "%d%s%02d", hours, res.getString(R.string.hour_separator), minutes);
        final String timeToStr = res.getString(hours == 0 ? R.string.to_m : R.string.to_h);

        return new String[]{timeStr, timeToStr};
    }

    public static long roundToInterval(long time) {

        int hours = (int) (time / Const.HOUR);
        int minutes = (int) ((time - hours * Const.HOUR) / Const.MINUTE);
        long newTime = 0;

        int index = minutes % 10;
        switch (index) {
            case 1:
            case 6:
                newTime -= Const.MINUTE;
                break;
            case 2:
            case 7:
                newTime -= 2 * Const.MINUTE;
                break;
            case 3:
            case 8:
                newTime += 2 * Const.MINUTE;
                break;
            case 4:
            case 9:
                newTime += Const.MINUTE;
                break;
        }

        return (hours * Const.HOUR) + (minutes * Const.MINUTE) + newTime;
    }

    public static int getCurrentProgressY(long timePast, long duration, int height, long maxTime) {
        if (duration > 0) {
            float timePercentage = (((float) timePast / (float) maxTime));
            return (int) (0.8f * height * timePercentage + (height * Const.INITIAL_PERCENTAGE));
        } else {
            return (int) (height * Const.INITIAL_PERCENTAGE);
        }
    }

    public static long getRoundedCurrentTimeMillis() {
        final Calendar now = Calendar.getInstance();
        now.set(Calendar.MILLISECOND, 0);
        now.set(Calendar.SECOND, 0);

        return now.getTimeInMillis();
    }

    public static long getNextRoundedMinuteTimeMillis() {
        return DateUtils.MINUTE_IN_MILLIS + getRoundedCurrentTimeMillis();
    }
}

File: PeaceOfMind+/app/src/main/java/ca/mudar/fairphone/peaceofmind/util/TimeHelper.kt
/*
 * Copyright (C) 2013 Mudar Noufal, PeaceOfMind+
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package ca.mudar.fairphone.peaceofmind.util

import android.content.Context
import android.text.format.DateFormat
import android.text.format.DateUtils
import ca.mudar.fairphone.peaceofmind.Const
import ca.mudar.fairphone.peaceofmind.R
import ca.mudar.fairphone.peaceofmind.model.AtPeaceRun
import ca.mudar.fairphone.peaceofmind.model.DisplayMode
import java.util.*
import kotlin.math.ceil
import kotlin.math.floor

object TimeHelper {
    private val HOUR_IN_MINUTES = 60
    private val NON_LEADING_ZERO_FORMAT = "%d"
    private val LEADING_ZERO_FORMAT = "%02d"

    fun hoursToSeekArcValue(hours: String): Int {
        return try {
            hoursToSeekArcValue(hours.toInt())
        } catch (e: NumberFormatException) {
            0
        }
    }

    fun hoursToSeekArcValue(hours: Int): Int {
        return hours * HOUR_IN_MINUTES / Const.SeekArc.GRANULARITY
    }

    fun getDurationForProgress(progress: Int,
                               @DisplayMode displayMode: String = DisplayMode._DEFAULT,
                               startTime: Long?): Long {
        fun getCeilDurationForStartTime(duration: Long, startTime: Long): Long {
            val roundStartTime = Const.Timer.END_TIME_ROUND *
                    ceil(startTime.toDouble() / Const.Timer.END_TIME_ROUND).toLong()

            return duration + (roundStartTime - startTime)
        }

        if (progress <= 0) {
            return 0
        }

        val duration = progressToMillis(progress)
        return when (displayMode) {
            DisplayMode.END_TIME -> getCeilDurationForStartTime(duration, getTimeWithoutSeconds(startTime))
            else -> duration
        }
    }

    fun getEndTimeForDuration(duration: Long?, startTime: Long?): Long {
        return getTimeWithoutSeconds(startTime) + (duration ?: 0)
    }

    fun getEndTimeLabel(context: Context, endTime: Long?): String {
        val millis = endTime ?: getTimeWithoutSeconds(Date().time)

        return DateFormat.getTimeFormat(context).format(millis)
    }

    fun getDurationLabel(context: Context, duration: Long?): String {
        val minutes = (duration ?: 0) / DateUtils.MINUTE_IN_MILLIS
        val hours = String.format(NON_LEADING_ZERO_FORMAT, minutes / HOUR_IN_MINUTES)
        val paddedMinutes = String.format(LEADING_ZERO_FORMAT, minutes % HOUR_IN_MINUTES)

        return context.resources.getString(R.string.duration_hours_minutes,
                hours, paddedMinutes)
    }

    private fun progressToMillis(progress: Int): Long {
        return when {
            (progress <= 0) -> 0L
            else -> progress * Const.SeekArc.GRANULARITY * DateUtils.MINUTE_IN_MILLIS
        }
    }

    private fun millisToProgress(millis: Long): Int {
        return when {
            (millis <= 0) -> 0
            else -> (millis / (Const.SeekArc.GRANULARITY * DateUtils.MINUTE_IN_MILLIS)).toInt()
        }
    }

    private fun getTimeWithoutSeconds(time: Long?): Long {
        val calendar = GregorianCalendar()
        time?.let {
            calendar.timeInMillis = time
        }

        // Set millis and seconds to zero
        calendar.set(Calendar.MILLISECOND, 0)
        calendar.set(Calendar.SECOND, 0)

        return calendar.timeInMillis
    }

    fun getProgressForAtPeaceRun(atPeaceRun: AtPeaceRun, @DisplayMode displayMode: String = DisplayMode._DEFAULT): Int {
        fun getFloorDuration(duration: Long): Long {
            return Const.Timer.END_TIME_ROUND *
                    floor(duration.toDouble() / Const.Timer.END_TIME_ROUND).toLong()
        }

        if (atPeaceRun.duration == null || atPeaceRun.duration <= 0) {
            return 0
        }

        val duration = when (displayMode) {
            DisplayMode.END_TIME -> getFloorDuration(atPeaceRun.duration)
            else -> atPeaceRun.duration
        }
        return millisToProgress(duration)
    }
}
